### PR TITLE
Windows Service Setup.exe

### PR DESCRIPTION
**Pull request by bencurthoys**
NSIS script to create Setup.exe for installing SyncThing on Windows as a
Service.

### CODE REVIEW

File: etc/windows/syncthingservice.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<service>
+  <id>SyncThing</id>
+  <name>SyncThing</name>
+  <description>This service runs SyncThing</description>
+  <executable>syncthing.exe</executable>
+  <arguments>-home .\ServiceHome -no-console -no-browser</arguments>
```
Comment (dzarda) [inline etc/windows/syncthingservice.xml]: So... What is the "-home" parameter set to? Current directory?

**Comment by calmh** [comment]
This looks neat, potentially something that we could build as part of the normal build and offer as a download. What's the result of all the build steps - an install.exe-file that installs the stuff somewhere and runs it as some user?

Perhaps the build instructions would be better in a `README.txt` or so, to be easier found?

Oh, and "SyncThing" -> "Syncthing" please.

**Comment by calmh** [comment]
Oh and remove the binary from the pull request, but I think I saw on IRC you were already working on that. :)

**Comment by bencurthoys** [comment]
On 06/03/2015 14:29, Jakob Borg wrote:
>
> This looks neat, potentially something that we could build as part of 
> the normal build and offer as a download. What's the result of all the 
> build steps - an install.exe-file that installs the stuff somewhere 
> and runs it as some user?
>

The service runs at the local system account by default.

> Perhaps the build instructions would be better in a |README.txt| or 
> so, to be easier found?
>
> Oh, and "SyncThing" -> "Syncthing" please.
>

Will do!

**Comment by bencurthoys** [comment]
On 06/03/2015 14:30, Jakob Borg wrote:
>
> Oh and remove the binary from the pull request, but I think I saw on 
> IRC you were already working on that. :)
>

So that binary is a small (58k) and basically static service wrapper, 
simply downloaded and then renamed.

It's from https://github.com/kohsuke/winsw
(which was mentioned in the comment thread 
https://discourse.syncthing.net/t/sycthing-as-a-service-on-windows/1934/3)

I'll take it out, not having it just adds a step to the build 
instructions =)

**Comment by kozec** [comment]
> The service runs at the local system account by default.

That's may have some probably non-desirable side effects as discussed on https://github.com/syncthing/syncthing/issues/906#issuecomment-62127943

**Comment by calmh** [comment]
What happens to files created by syncthing when running like that, assuming it's syncing into a directory owned by "me"? Do those files inherit all the relevant permissions so they look "mine" by default?

**Comment by bencurthoys** [comment]
On 06/03/2015 14:38, kozec wrote:
>
>     The service runs at the local system account by default.
>
> That's may have some probably non-desirable side effects as discussed 
> on #906 (comment) 
> <https://github.com/syncthing/syncthing/issues/906#issuecomment-62127943>
>

Just reading down the thread... if it's the home directory location 
you're referring to - that's why the service wrapper xml config sets 
-home to be the install path + "ServiceHome".

Or were there other issues?

If there were an installer, I would expect it to copy syncthing.exe to 
C:\Program Files\Syncthing, and then optionally install the service.

 > What happens to files created by syncthing when running like that, 
assuming it's syncing into a directory owned by "me"? Do those files 
inherit all the relevant permissions so they look "mine" by default?

I'm just looking at it on my machine, where it's running as local 
system, and my user account (which owns the containing folder) also owns 
the created files.

File ownership is sort of less important on Windows than on Linux. The 
only time it's ever really an issue is for files inside people's 
profiles, which running as a service would explcitly not touch. The 
point of running as a service is that it runs even when no one's logged 
in, which is handy for some use cases (like the stuff I want to do!) and 
irrelevant for others =)

**Comment by AudriusButkevicius** [comment]
Can you squash it into a single commit?

**Comment by bencurthoys** [comment]
Um. I'll have to look up how to do that!

**Comment by AudriusButkevicius** [comment]
```
git rebase -i HEAD^^^^
```
replace all but one `edit` with `squash`
then
```
git push origin master -f
```

**Comment by bencurthoys** [comment]
It says "Sucessfully rebased and updates refs/heads/master", and then opens a text file. I edit the text file, changing "pick" to "squash". Then I run the push, and it says "Everything is up to date". 

Sorry, sure you've got better things to be doing that walking me through Git....

Ah - got it! When it opens the text editor, it waits for the process to terminate. I'm using notepad++, and was just opening and closing tabs, so it wasn't triggering the next bit when I closed the tab because the process was still running. It's now rebasing 2/4...

**Comment by bencurthoys** [comment]
There.

**Comment by AudriusButkevicius** [comment]
LGTM! 
I wonder if we can get the Windows Jenkins slave to do this dance for us, so that we could include it as part of the official releases.

**Comment by calmh** [comment]
> I wonder if we can get the Windows Jenkins slave to do this dance for us, so that we could include it as part of the official releases (which are cross-compiled to Windows on Linux currently).

This would be nice...

**Comment by StefanScherer** [comment]
I just wanted to suggest to use an appveyor.yml, but now I see that there is one already. Good starting point.

Any help needed here?
I have done some AppVeyor + nssm + gobinary MSI Builds in the last few weeks, which is quite easy, but I do not have experience with NSIS.

**Comment by AudriusButkevicius** [comment]
I think it's just dowloading nsis, and compiling the script.

**Comment by StefanScherer** [comment]
OK, I could try it in a fork. AppVeyor already has chocolatey installed, so a `choco install -y nsis` could do it...

**Comment by StefanScherer** [comment]
So, this is my work in progress. AppVeyor is able to build a `SyncThingSetup.exe`.

My working branch is here https://github.com/StefanScherer/syncthing/tree/build-setup-exe-with-appveyor and the two major files needed are:
* [appveyor.yml](https://github.com/StefanScherer/syncthing/blob/build-setup-exe-with-appveyor/appveyor.yml)
* [etc/windows/build_setup_exe.ps1](https://github.com/StefanScherer/syncthing/blob/build-setup-exe-with-appveyor/etc/windows/build_setup_exe.ps1)

The first artifact can be found on AppVeyor at:
https://ci.appveyor.com/project/StefanScherer/syncthing/build/build-setup-exe-with-appveyor-17/artifacts

But I'm not quite happy with it. This is still a work in progress as there are still questions:
* build a x86 and amd64 setup.exe or NSIS has to support both CPU architectures
* the target directory of the amd64 binary shouldn't be `C:\Program Files (x86)\Syncthing`
* The exe files and the installer should be code signed. Otherwise Windows will complain about unkwown  publisher.
* Some part of the installtion needs .NET 3.5 which isn't installed on Windows 10. This seems to be a dependency of winsx. So I suggest using nssm.exe which has no other dependencies.

**Pull request merged**